runtime/src/installed_scheduler_pool.rs

```diff
@@ -0,0 +1,446 @@
+//! Transaction processing glue code, mainly consisting of Object-safe traits
+//!
+//! `trait InstalledSchedulerPool` is the most crucial piece of code for this whole integration.
+//!
+//! It lends one of pooled `trait InstalledScheduler`s out to a `Bank`, so that the ubiquitous
+//! `Arc<Bank>` can conveniently work as a facade for transaction scheduling, to higher-layer
+//! subsystems (i.e. both `ReplayStage` and `BankingStage`). `BankForks`/`BankWithScheduler` and
+//! some `Bank` methods are responsible for this book-keeping, including returning the scheduler
+//! from the bank to the pool after use.
+//!
+//! `trait InstalledScheduler` can be fed with `SanitizedTransaction`s. Then, it schedules and
+//! commits those transaction execution results into the associated _bank_. That means,
+//! `InstalledScheduler` and `Bank` are mutually linked to each other, resulting in somewhat
+//! special handling as part of their life-cycle.
+//!
+//! Albeit being at this abstract interface level, it's generally assumed that each
+//! `InstalledScheduler` is backed by multiple threads for performant transaction execution and
+//! there're multiple independent schedulers inside a single instance of `InstalledSchedulerPool`.
```

Comment:
the current situation is complicated.

in short, this is always the case. as soon as a new fork is created by `BankForks::insert()`, a new `PooledScheduler` is spawn, which means a new independent group of threads from other ongoing forks (sharing no synchronization primitives among others). This is true, regardless `--replay-slots-concurrently` or not.


So, _when `--replay-slots-concurrently` is enabled_, unified scheduler can take advantage of it in that _it can fully concurrently replay ledger entries (note: by principle!)_. on the other hand, blockstore_processor uses a shared rayon thread group. While the thread group is populated with N (core count) threads, it's still possible for any blocking call there to starve the other fork. also, I've observed non-optimal rayon behavior when `.install()` is called from multiple threads against a shared thread group.

the caveat for the full concurrency of unified scheduler is that it's true _if there's no single `is_complete()` bank at given moment among all of ongoing forks_. that's because `process_replay_results` isn't concurrent regarding multiple forks even if `--replay-slots-concurrently` is enabled (still). so, as soon as any buffered `schedule_execution()`-ed transactions are all executed, other `!is_complete()` forks can be starved sub-optimally.

Anyway, these are all edge case runtime behavior...  I'll planning to revisit this in distant future... or @bw-solana might be planning to further enhance the replay stage's concurrency. xD

Reply:
(while writing the above, I noticed i better off pulling a small assert from the dirty topic branch: 4dbbcbdf7ec2292d5baee3cb69c3d9c9698886be, pending ci for its correctness... lol)

Reply:
oh, this concurrency bla bla also affected by this pending feature activation: https://github.com/solana-labs/solana/pull/31239/files#r1174796706